Report worker lost storage information in doctor storage CLI

LuQQiu opened the pull request:
Report worker lost storage.

```
$./bin/alluxio fsadmin doctor storage
No worker storage errors or warnings.
```

or if some of the storage cannot be initialized or is lost during periodically storage check
```
$./bin/alluxio fsadmin doctor storage
The following directories are lost in worker AlluxioSandbox-workers-1:
/ssdone/alluxioWorker (SSD)
/ssdtwo/alluxioWorker (SSD)
```

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2794/
Test PASSed.

LuQQiu commented:
@bf8086 Added the `fsadmin doctor storage` command, the main logics is finished and tested. Now is working on adding more tests, PTAL, thanks

bf8086 commented on shell/src/main/java/alluxio/cli/fsadmin/doctor/StorageCommand.java:
```diff
@@ -0,0 +1,62 @@
+    List<WorkerLostStorageInfo> workerLostStorageInfo = mBlockMasterClient.getWorkerLostStorage();
+    for (WorkerLostStorageInfo info : workerLostStorageInfo) {
+      Map<String, List<String>> lostStorageOnTiers = info.getLostStorageOnTiers();
+      if (lostStorageOnTiers.size() != 0) {
```
Can we add a message to let user know if all worker storage paths are in working state?

LuQQiu replied:
Added the message

bf8086 commented on core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java:
```diff
@@ -831,7 +845,9 @@ public void workerRegister(long workerId, List<String> storageTiers,
       Map<String, Long> usedBytesOnTiers, List<Long> removedBlockIds,
-      Map<String, List<Long>> addedBlocksOnTiers, List<Metric> metrics) {
+      Map<String, List<Long>> addedBlocksOnTiers,
+      Map<String, List<String>> removedStorageOnTiers,
```
Does the command work after master is restarted? We might need to report all existing lost storage paths when worker is registered with a new master node.

LuQQiu replied:
Previously we record failed to initialize paths and report via worker register and report failed in storage check paths in worker heartbeat which cannot survive through restart.

Now we record all the lost storage paths(failed to initialize and failed in periodical storage check) in the worker and report them in the worker register. When lost storage find in storage check, we report them via worker heartbeat and also record them in the worker to make this command work after master is restarted.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2856/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2858/
Test PASSed.

LuQQiu commented on core/transport/src/grpc/block_master.proto:
```diff
@@ -98,6 +98,16 @@ message GetWorkerInfoListPResponse {
 
+message WorkerLostStorageInfo {
+  optional grpc.WorkerNetAddress address = 1;
+  map<string, StorageList> lostStorageOnTiers = 2;
```
Added the comments, thanks

LuQQiu commented on core/transport/src/grpc/block_master.proto:
```diff
@@ -132,12 +142,21 @@ service BlockMasterClientService {
 }
 
+message StorageList {
+  repeated string storage = 1;
```
Add the comments for `StorageList` and `storage`

LuQQiu commented on core/server/worker/src/main/java/alluxio/worker/block/BlockHeartbeatReport.java:
```diff
@@ -55,4 +63,14 @@ public BlockHeartbeatReport(Map<String, List<Long>> addedBlocks, List<Long> remo
+   * Gets the directory paths which were removed from this worker
+   * in the heartbeat this report represents.
+   *
+   * @return a map from storage tier alias to removed storage
```
Changed to use `storage paths` instead of `storage` in the implementations

LuQQiu commented on core/common/src/main/java/alluxio/wire/WorkerLostStorageInfo.java:
```diff
@@ -0,0 +1,63 @@
+/**
+ * Class to represent a ufs info.
+ */
+public final class WorkerLostStorageInfo implements Serializable {
```
Removed this wire type and use the grpc type in all the places that need to use it.

LuQQiu commented on core/server/master/src/main/java/alluxio/master/block/BlockMaster.java:
```diff
@@ -83,6 +84,13 @@
       throws UnavailableException, InvalidArgumentException;
 
+  /**
+   * Gets the list of worker lost storage information.
```
Removed this line and directly use the return information

LuQQiu commented on core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java:
```diff
@@ -973,10 +975,30 @@ public boolean checkStorage() {
+  /**
+   * Processes the failed to initialize storage.
+   *
+   * @param failedStorage the failed to initialize storage to process
```
Done

LuQQiu commented on core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java:
```diff
@@ -973,10 +975,30 @@ public boolean checkStorage() {
+   *
+   * @param failedStorage the failed to initialize storage to process
+   */
+  private void processFailedToInitializeStorage(Map<String, List<String>> failedStorage) {
```
Removed

LuQQiu commented on core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java:
```diff
@@ -470,6 +471,19 @@ public long getUsedBytes() {
+      synchronized (worker) {
+        workerLostStorageList.add(new WorkerLostStorageInfo()
+            .setWorkerAddress(worker.getWorkerAddress())
+            .setLostStorageOnTier(worker.getRemovedStorages()));
```
Use `LostStorage` In all the place instead of multiple naming. Thanks

LuQQiu commented:
@bf8086 Addressed the comments. Now this command is able to work through master restart and gain/lost leadership. Still working on the tests. PTAL, thanks

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2919/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2923/
Test FAILed.

LuQQiu commented:
jenkins, test this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2935/
Test PASSed.

LuQQiu commented:
@bf8086 Added the related tests and support report lost storage across master restart. PTAL, thanks

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2977/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2979/
Test FAILed.

bf8086 commented on core/server/worker/src/main/java/alluxio/worker/block/BlockMasterSync.java:
```diff
@@ -107,7 +107,8 @@ private void registerWithMaster() throws IOException {
         storageTierAssoc.getOrderedStorageAliases(), storeMeta.getCapacityBytesOnTiers(),
-        storeMeta.getUsedBytesOnTiers(), storeMeta.getBlockList(), configList);
+        storeMeta.getUsedBytesOnTiers(), storeMeta.getBlockList(),
+        storeMeta.getLostStorage(), configList);
```
Will this parameter include all lost storage paths including the new ones reported in heartbeat with the previous active master?

LuQQiu replied:
The `LostStorageIntegrationTest.lostStorageWhenRestart()` is designed to make sure that the lost storage information in the previous master can sustain through master restarts. This parameter include all lost storage paths that this worker failed to initialize or lost (which are reported to the previous masters)

bf8086 replied:
I see. Thanks for the explanation.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2982/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2983/
Test PASSed.

LuQQiu commented on core/server/master/src/main/java/alluxio/master/block/BlockMasterWorkerServiceHandler.java:
```diff
@@ -73,10 +73,10 @@ public void blockHeartbeat(BlockHeartbeatPRequest request,
+    final Map<String, StorageList> lostStorage = request.getLostStorageMap();
+    Map<String, List<String>> lostStorageMap = new HashMap<>();
+    for (Map.Entry<String, StorageList> storageEntry : lostStorage.entrySet()) {
+      lostStorageMap.put(storageEntry.getKey(), storageEntry.getValue().getStorageList());
```
Thanks for the suggestions! Changed the places that I need to do map transformation to use lambda expressions

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2984/
Test PASSed.

LuQQiu commented:
Thanks @bf8086 
@aaudiber In this PR, every worker will 
(1) initialize the storage paths. If failed -> record lost storage paths in worker and report to master in worker register
(2) do periodically storage paths check. If failed 
     -> report to master in worker heartbeat. & record lost storage paths in worker

So in worker register, the master is able to know all the previous lost storage paths no matter they are failed to initialize or are found lost in the periodical storage check.

In worker heartbeat, the master is able to know the storage lost in the last heartbeat interval.

In the master side, we record the lost storage information in the `MasterWorkerInfo` of workers. When clients issue a `getWorkerLostStorage` request, we provide them that information.

PTAL, thanks

aaudiber commented on core/server/worker/src/main/java/alluxio/worker/block/meta/StorageTier.java:
```diff
@@ -227,6 +231,13 @@ public StorageDir getDir(int dirIndex) {
+   * @return a list of lost storage paths
+   */
+  public List<String> getLostStorage() {
+    return Collections.unmodifiableList(mLostStorage);
```
seems susceptible to concurrent modification exceptions. Could we return a copy instead?

LuQQiu replied:
Changed to return a copy, thanks

aaudiber commented on core/transport/src/grpc/block_master.proto:
```diff
@@ -98,6 +98,17 @@ message GetWorkerInfoListPResponse {
+message WorkerLostStorageInfo {
+  optional grpc.WorkerNetAddress address = 1;
+  /** a map from tier alias to the lost storage paths */
+  map<string, StorageList> lostStorage = 2;
```
`optional`

LuQQiu replied:
Grpc is different from Netty that `Field labels (required/optional/repeated) are not allowed on map fields.`, thanks

LuQQiu commented on core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockStoreMeta.java:
```diff
@@ -53,6 +53,9 @@
 
   private final StorageTierAssoc mStorageTierAssoc;
 
+  /** Mapping from tier alias to lost storage paths. */
```
Yeah, but in our code base, we usually use tier alias referring to storage dir tier.

LuQQiu commented on core/server/worker/src/main/java/alluxio/worker/block/BlockHeartbeatReporter.java:
```diff
@@ -58,10 +65,12 @@ public BlockHeartbeatReport generateReport() {
       // Copy added and removed blocks
       Map<String, List<Long>> addedBlocks = new HashMap<>(mAddedBlocks);
       List<Long> removedBlocks = new ArrayList<>(mRemovedBlocks);
+      Map<String, List<String>> removedStorage = new HashMap<>(mLostStorage);
```
The copy should be in the `BlockHeartbeatReport` but the logics that we need to copy it first and then clear the map make it easier to locate in the `BlockHeartbeatReporter`

LuQQiu commented on core/server/master/src/main/java/alluxio/master/block/meta/MasterWorkerInfo.java:
```diff
@@ -175,6 +178,31 @@ public void removeBlock(long blockId) {
+   */
+  public void addLostStorage(Map<String, List<String>> lostStorage) {
+    for (Map.Entry<String, List<String>> entry : lostStorage.entrySet()) {
+      List<String> paths = mLostStorage.getOrDefault(entry.getKey(), new ArrayList<>());
```
Here will reduce many map get and put operations because we will `paths.addAll(List<String>)` instead of one by one `paths.add(String storage` in the `addLostStorage(tierAlias, dirPath`

LuQQiu commented on core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockStoreMeta.java:
```diff
@@ -163,6 +171,15 @@ protected DefaultBlockStoreMeta(BlockMetadataManager manager, boolean shouldIncl
     }
+
+    for (StorageTier tier : manager.getTiers()) {
+      if (tier.getLostStorage().size() != 0) {
```
Changed, thanks

LuQQiu commented on core/server/master/src/main/java/alluxio/master/block/meta/MasterWorkerInfo.java:
```diff
@@ -330,11 +358,26 @@ public boolean isRegistered() {
+   * @return the map from tier alias to lost storage paths in this worker
+   */
+  public Map<String, List<String>> getLostStorage() {
+    return Collections.unmodifiableMap(mLostStorage);
```
Changed to use a copy. Oh, I first thought `Collections.unmodifiableMap()` is making a read-only copy. Thanks for the explanation.

LuQQiu commented on core/server/master/src/main/java/alluxio/master/block/meta/MasterWorkerInfo.java:
```diff
@@ -175,6 +178,31 @@ public void removeBlock(long blockId) {
   }
 
+  /**
+   * Adds a new worker lost storage paths.
```
Changed, thanks

LuQQiu commented on core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java:
```diff
@@ -846,6 +870,10 @@ public Command workerHeartbeat(long workerId, Map<String, Long> capacityBytesOnT
       processWorkerAddedBlocks(worker, addedBlocksOnTiers);
       processWorkerMetrics(worker.getWorkerAddress().getHost(), metrics);
 
+      if (lostStorage != null) {
```
It will never be null in actual server running, remove the `null` check, thanks

LuQQiu commented on core/server/master/src/main/java/alluxio/master/block/BlockMasterWorkerServiceHandler.java:
```diff
@@ -67,20 +64,23 @@ public void blockHeartbeat(BlockHeartbeatPRequest request,
+        .entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey,
+            e -> e.getValue().getTiersList()));
+
+    final Map<String, List<String>> lostStorageMap = request.getLostStorageMap()
```
Use `StorageList` when we only have read needs. If the list need to add new elements, we use `List<String>`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2993/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2994/
Test PASSed.

LuQQiu commented:
@aaudiber Thanks for the review. Addressed the comments, PTAL

aaudiber commented on core/server/worker/src/main/java/alluxio/worker/block/meta/StorageTier.java:
```diff
@@ -235,7 +235,7 @@ public StorageDir getDir(int dirIndex) {
    */
   public List<String> getLostStorage() {
-    return Collections.unmodifiableList(mLostStorage);
+    return new ArrayList<>(mLostStorage);
```
can we update `getStorageDirs()` to also return a copy? It has the same issue

LuQQiu replied:
Now the other places except the `BlockHeartbeatReport` all returns a copy of `lost storages`

LuQQiu commented on core/server/worker/src/main/java/alluxio/worker/block/BlockHeartbeatReport.java:
```diff
@@ -55,4 +63,14 @@ public BlockHeartbeatReport(Map<String, List<Long>> addedBlocks, List<Long> remo
+   * @return a map from storage tier alias to lost storage paths
+   */
+  public Map<String, List<String>> getLostStorage() {
+    return Collections.unmodifiableMap(mLostStorage);
```
@aaudiber Now only here I use the `unmodifableMap` because the map in the `blockheartbeatReport` is fixed and will never be updated.

aaudiber replied:
It's valuable to be able to enforce properties within a single file. Right now the parameters passed to the constructor happen to not be changed externally, but someone could change that in the future without looking at this class. If we copy the parameters in the constructor, this class won't have concurrency issues even if the lists are changed by code outside this class.

LuQQiu replied:
Make sense, now I move the copy process into the `BlockHeartbeatReport` to confirm that the maps and lists inside `BlockHeartbeatReport` are never modified. Thanks

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/3032/
Test PASSed.

aaudiber commented:
alluxio-bot, merge this please
